Improve install, usage, and configuration guide (#2677)
nf-core/website: sites/docs/src/content/docs/usage/configuration/introduction.md

---
title: Configuration
subtitle: How configure nf-core pipelines
shortTitle: Configuration options
weight: 1
parentWeight: 20
---

## Configure nf-core pipelines

> Each nf-core pipeline comes with a set of “sensible defaults”.

Reviewer: In regards to waht exactly? Memory resources? Parameters?

Reviewer: I think it's important what to use them for

Author: Now framed “sensible defaults” for a typical analysis of full size data. Will expand below

While the defaults are a great place to start, you will certainly want to modify these to fit your own purposes and system requirements.

When a pipeline is launched, Nextflow will look for config files in several locations.
As each source can contain conflicting settings, the sources are ranked to decide which settings to apply.

nf-core pipelines may utilize any of these configuration files.

Configuration sources are reported below and listed in order of priority:

> 1. Parameters specified on the command line (`--parameter`)

Reviewer: I think we should drop this list to rather point towards the Nextflow docs, unless this changes

Author: I see this as minimal framing for the rest of the section. E.g., how nextflow.config and parameter files fit together.

2. Parameters that are provided using the `-params-file` option
3. Config file that are provided using the `-c` option
4. The config file named `nextflow.config` in the current directory
5. The config file named `nextflow.config` in the pipeline project directory
6. The config file `$HOME/.nextflow/config`
7. Values defined within the pipeline script itself (e.g., `main.nf`)

While some of these files are already included in the nf-core pipeline repository (e.g., the `nextflow.config` file in the nf-core pipeline repository), some are automatically identified on your local system (e.g., the `nextflow.config` in the launch directory), and others are only included if they are specified using run options (e.g., `-params-file`, and `-c`).

> If you clone and manually edit an nf-core pipeline then you cannot update to more recent versions of the pipeline without overwriting your changes. You also risk moving away from the canonical pipeline and losing reproducibility.

Reviewer: I would be stronger about this, say you SHOULD NOT manually edit an nf-core pipeline downloaded by nextflow

:::tip

> Configuration locations can be used to your benefit. For example, use `-profile` for your cluster configuration, `~/.nextflow/config` for your personal configuration, and `nextflow.config` in your working directory for run-specific configuration.

Reviewer: Or maybe `-c` for the last one? Then you can name it with run specific name

:::

### Parameters

Parameters are pipeline specific settings that can be used to customize the execution of a pipeline.

At the highest level, parameters can be customized using the command line. Any parameter can be configured on the command line by prefixing the parameter name with a double dash (--):

```bash
--<parameter>
```

Depending on the parameter type, you may be required to add additional information after your parameter flag.
For example, you would add string parameter after the parameter flag for the `nf-core/rnaseq` `--input` and `--output` parameters.

```bash

> nextflow nf-core/rnaseq --input <path/to/input> --outdir <results>

Reviewer: ```suggestion
nextflow nf-core/rnaseq --input <path/to/input> --outdir <path/to/results>
```

```

Every nf-core pipeline has a full list of parameters on the nf-core website. You will be shown a description and the type of the parameter when viewing these parameters. Some parameters will also have additional text to help you understand how a parameter should be used. See the [parameters page of the nf-core rnaseq pipeline](https://nf-co.re/rnaseq/3.14.0/parameters/).

### Default configuration files

All parameters have a default configuration that is defined using the `nextflow.config` file in the root of the pipeline directory. Many parameters are set to `null` or `false` by default and are only activated by a profile or config file.

> There are also several `includeConfig` statements in the `nextflow.config` file that are used to include additional config files from the `conf/` folder. Each additional `.config` file contains categorized configuration information for your pipeline execution, some of which can be optionally included as profiles:

Reviewer: ```suggestion
nf-core pipelines also include additional config files from the `conf/` folder of a pipeline repository. Each additional `.config` file contains categorized configuration information for your pipeline execution, some of which can be optionally included as profiles:
```

- `base.config`
  - Included by the pipeline by default
  - Generous resource allocations using labels
  - Does not specify any method for software dependencies and expects software to be available (or specified elsewhere)
- `igenomes.config`
  - Included by the pipeline by default
  - Default configuration to access reference files stored on AWS iGenomes
- `modules.config`
  - Included by the pipeline by default
  - Module-specific configuration options (both mandatory and optional)
- `test.config`
  - Only included if specified as a profile
  - A configuration profile to test the pipeline with a small test dataset
- `test_full.config`
  - Only included if specified as a profile

>   - A configuration profile to test the pipeline with a full-size test dataset

Reviewer: I guess we are getting a bit more advanced than the quick_Start bit, so I would recommend again saying that 'you should not modify these'

Author: Added below

:::note
Some configuration files contain the definition of profiles that can be flexibly applied. For example, the `docker`, `singularity`, and `conda` profiles are defined in the `nextflow.config` file in the pipeline project directory.
:::

Profiles are sets of configuration options that can be flexibly applied to a pipeline.
They are also commonly defined in the `nextflow.config` file in the root of the pipeline directory.

> Profiles used by nf-core pipelines can be broadly categorized into two groups:

Reviewer: ```suggestion
Profiles that come with nf-core pipelines can be broadly categorized into two groups:
```

- Software management profiles
  - Profiles for the management of software dependencies using container or environment management tools, for example, `docker`, `singularity`, and `conda`.
- Test profiles
  - Profiles to execute the pipeline with a standardized set of test data and parameters, for example, `test` and `test_full`.

nf-core pipelines are required to define software containers and environments that can be activated using profiles. Although it is possible to run the pipelines with software installed by other methods (e.g., environment modules or manual installation), using container technology is more sharable, convenient, and reproducible.

> ### Shared configuration files

Reviewer: Iwonder if this should berenamed to 'shared infrastructure files' or something similar, as that is waht they are used for, not som uch sharing e.g. pipeline parameters between users

Author: I think the focus here is the source of configuration rather than the bias towards these being used for infrastruture.

> An `includeConfig` statement in the `nextflow.config` file is also used to include custom institutional profiles that have been submitted to the [nf-core config repository](https://github.com/nf-core/configs). At run time, nf-core pipelines will fetch these configuration profiles from the [nf-core config repository](https://github.com/nf-core/configs) and make them available.

Reviewer: ```suggestion
nf-core pipelines can also load custom institutional profiles that have been submitted to the [nf-core config repository](https://github.com/nf-core/configs). At run time, nf-core pipelines will fetch these configuration profiles from the [nf-core config repository](https://github.com/nf-core/configs) and make them available.
```

Follow [this tutorial](https://nf-co.re/docs/usage/tutorials/step_by_step_institutional_profile) to set up your own institutional profile.

### Custom parameter and configuration files

Nextflow will look for files that are external to the pipeline project directory. These files include:

- The config file `$HOME/.nextflow/config`
- A config file named `nextflow.config` in your current directory
- Custom configuration files specified using the command line
  - A parameter file that is provided using the `-params-file` option
  - A config file that are provided using the `-c` option

**Parameter file format**

Parameter files are `.json` files that can contain an unlimited number of parameters:

```json title="nf-params.json"
{
  "<parameter1_name>": 1,
  "<parameter2_name>": "<string>",
  "<parameter3_name>": true
}
```

You can override default parameters by creating a `.json` file and passing it as a command-line argument using the `-param-file` option.

```bash
nextflow run nf-core/rnaseq -profile docker --input <path/to/input? --outdir <results> -param-file <path/to/nf-params.json>
```

> **Configuration file format**

Reviewer: If we are talking specifically about pipeline parameters, I would remove this entire section as this is not recommended by the Nextflow developers apparently - we should stick just to params json/yamls

Author: This section is more biased towards describing the file formats.

I'm lumping everything that is configuration realted together to mirror the nextflow docs. This get's complicated as there are params in the test config files as well. I'll add another warning elsewhere to highlight this again.

Configuration files are `.config` files that can contain various pipeline properties and can be passed to Nextflow using the `-c` option in your execution command:

```bash
nextflow run nf-core/rnaseq  -profile docker --input <path/to/input> --outdir <results> -c <path/to/custom.config>
```

Custom configuration files are the same format as the configuration file included in the pipeline directory.

Configuration properties are organized into scopes by dot prefixing the property names with a scope identifier or grouping the properties in the same scope using the curly brackets notation. For example:

```groovy
alpha.x  = 1
alpha.y  = 'string value'
```

Is equivalent to:

```groovy
alpha {
    x = 1
    y = 'string value'
}
```

### Scopes

Scopes allow you to quickly configure settings required to deploy a pipeline on different infrastructure using different software management.

A common scenario is for users to write a custom configuration file specific to running a pipeline on their infrastructure.

> :::warning

Reviewer: Ah this should go above, but I have a feeling talking ab out `scopes` here might be getting into too much detail, maybe drop this whole section to say 'to see all options for how to configure your pipeline see the nextflow docs'

Author: I think it needs to be included, and I agree it is potentially too much detail, but I think it's important to frame that they exist, can be added to the same config file, and address different configuration needs.

I try to stay away from params here to prevent confusion.

Author: I've merged this into the config file descritpion to try to improve the flow.
